### DISCUSSION TITLE
PowerOnState switch dependant

### DISCUSSION THREAD

**Discussion by assafkolin** [arendst/Tasmota#15029] [Q&A] [answered]
When there is a short circuit in the electrical device connected to the relay, a circuit breaker turned off automatically. The problem is that in this case it can't be turned on, because Tasmota maintained its last on state and whenever get power back, it retain the short circuit and the circuit breaker acts (and the device can't be located unlike regular switch which turned off before power on again).
It should be able to set PowerOnState to the last state given the last switch state.
If the switch attached to the Tasmota changes its state during power outage, and then when the Tasmota power comes back it won't turn on the device and the corrupted device can be located.

**Comment by ascillato2** [comment]
Hi, this can be achieved by the use of rules like in the following example:

```JS
PowerOnState 0
SwitchMode1 1
Rule1 1
Rule1 ON Switch1#Boot DO Power1 %value% ENDON
```

The rule will "follow" the state of the switch at boot time no matter the previous state.

**Reply by assafkolin** [reply]
Hi, thanks for the answer, but I'm not sure it solves my problem.
What I want is: When tasmota boots, If Switch1 changes it state from last saved Swich1 state (e.g. the switch was toggeled while the tasmota was off. Either from on to off, or off to on), then toggle Power1 from its last saved state (again: off to on, or on to off).

BTW, I'm not sure if I understand your rule correctly: it turns power on whrn the switch is on at boot?

**Reply by barbudor** [reply]
It assume that you are using Switchmode 1 (follow mode) and will sync the Power to the state of the switch at boot

**Reply by assafkolin** [reply]
No,
I use the default SwitchMode 0 mode. So both the physical switch and the software (hass) swich toggles. What I want is to be able to toggle even when tasmota is off wuthout electricity (so, in case of a short i will be able to toggles it to off, befaure electricity returns), without change the regular toggle behavior (of both software and hardware switches).

**Reply by ascillato** [reply]
If you are using toggle in a rocker switch, you can change `switchmode` to `1` (follow) and it will continue working as previously (for your case), plus that rule will do what you want.

Please, test it.

_(Rule updated with the needed `switchmode` precondition - Thanks @barbudor)_

**Reply by barbudor** [reply]
If you problem is only in case of a short-circuit, why doesn't `PowerOnState 0` suits your need ?
It will force Tasmota to always restart with Power OFF after a power-cut (but not for a "simple" software restart)

**Reply by assafkolin** [reply]
Because in this case after every blackout (which happens a lot in my town) the whole house will be dark until I'll turn on every light manually.

**Reply by barbudor** [reply]
How do you expect Tasmota to differentiate between the power grid down and a breaker cutting power because of a short circuit ?
For Tasmota, this is all the same

**Reply by ascillato** [reply]
> Because in this case after every blackout (which happens a lot in my town) the whole house will be dark until I'll turn on every light manually.

No, with the above rule, your device will just copy the state of your switch. So, if your light was ON, and you turn it OFF when there is no power, it will start at OFF when the power comes back. And also, if your light was OFF and you move your switch to ON position, it will be ON when the electricity comes back.

**Reply by assafkolin** [reply]
> How do you expect Tasmota to differentiate between the power grid down and a breaker cutting power because of a short circuit ? For Tasmota, this is all the same

I don't expect it. When there is short curcuit, I'll figure it out and I'll toggle the hardware switch before turning on the breaker. In this method I'll will be able to locate and isolate the shorted device (and fix it)

**Reply by barbudor** [reply]
So follow mode and Adrian's rule is the only way

**Reply by assafkolin** [reply]
> > Because in this case after every blackout (which happens a lot in my town) the whole house will be dark until I'll turn on every light manually.
> 
> No, with the above rule, your device will just copy the state of your switch. So, if your light was ON, and you turn it OFF when there is no power, it will start at OFF when the power comes back. And also, if your light was OFF and you move your switch to ON position, it will be ON when the electricity comes back.

Thanks, I'll try it. Just to make sure: will it works on the opposite case: if the light was ON, and the swich was OFF before, and I turn it ON (toggles it) when there is no power, it will start at OFF when the power comes back?

**Reply by barbudor** [reply]
Follow means that it will follow the state of the switch 
Switch is ON => Power ON
Switch is OFF => Power OFF

**Comment by ascillato2** [comment, marked as answer]
Another approach could be this one:

- One GPIO is defined as RELAY 1: Your light.
- Another GPIO is defined as SWITCH 1: Your rocker switch _(by rocker switch I meant a flip switch not a pushbutton)_.
- MEM1 is the state of your SWITCH 1.

Console commands:

```JS
PowerOnState 3
SwitchMode1 1
SwitchTopic 0
Mem1 0
Rule1 1
Rule1 ON Switch1#State DO MEM1 %value% ENDON ON Switch1#Boot!=%mem1% DO Power1 Toggle ENDON
```

The rule will toggle the light at boot time, ONLY if the state of the switch have changed since the previous saved state.

Seems that this new configuration does what you want:
* If Switch1 is in the same state than before => Do nothing
* If Switch1 is in a different state than before => Toggle

Anyway, I prefer the [previous rule](https://github.com/arendst/Tasmota/discussions/15029#discussioncomment-2317744) because anyone can define exactly the desired state of light as Tasmota will follow the switch no matter its previous state:
* Switch is ON => Power ON
* Switch is OFF => Power OFF

**Reply by ascillato2** [reply]
If you are doing this for safety, you may change in the second rule the last part from `Power1 Toggle` to `Power1 Off`, so:

- If Switch1 is in the same state than before => Do nothing
- If Switch1 is in a different state than before => Light Off

**Reply by assafkolin** [reply]
Thanks. That seemed to be the right direction. But i intend to use SwitchMode1 0, so I can use the software switch as well (using homeassistant, when everything is normal of course). Will your rule will also work in this SwitchMode1?

**Reply by ascillato** [reply]
With Switchmode1 0 (toggle for pushbuttons) Tasmota can't know the previous state because it is just watching for the transition 0 to 1 and not the state. For checking the actual state of the flip switch, you need Switchmode1 1.

For using it with Home Assistant, you can use it with switchmode1 1, no problem with that. I don't understand why you want to use a Flip switch as a Pushbutton and how you are using it in Home Assistant.

**Reply by assafkolin** [reply]
> Another approach could be this one:
> 
> * One GPIO is defined as RELAY 1: Your light.
> * Another GPIO is defined as SWITCH 1: Your rocker switch _(by rocker switch I meant a flip switch not a pushbutton)_.
> * MEM1 is the state of your SWITCH 1.
> 
> Console commands:
> 
> ```js
> PowerOnState 3
> SwitchMode1 1
> SwitchTopic 0
> Mem1 0
> Rule1 1
> Rule1 ON Switch1#State DO MEM1 %value% ENDON ON Switch1#Boot!=%mem1% DO Power1 Toggle ENDON
> ```
> 
> The rule will toggle the light at boot time, ONLY if the state of the switch have changed since the previous saved state.
> 
> Seems that this new configuration does what you want:
> 
> * If Switch1 is in the same state than before => Do nothing
> * If Switch1 is in a different state than before => Toggle
> 
> Anyway, I prefer the [previous rule](https://github.com/arendst/Tasmota/discussions/15029#discussioncomment-2317744) because anyone can define exactly the desired state of light as Tasmota will follow the switch no matter its previous state:
> 
> * Switch is ON => Power ON
> * Switch is OFF => Power OFF

I finally managed to do the desired task and test it successfully. 
I slightly modified [ascillato2](https://github.com/ascillato2) suggestion so there will be "SwitchMode1 0" behavior when the configuration is "SwitchMode1 1" (so the Switch1#State value can be used and won't be constantly 2 as in toggle mode).

the console commands are then:

PowerOnState 3
SwitchMode1 1
SwitchTopic 0
Mem1 0
Rule1 1
Rule1 ON Switch1#State DO Backlog Power1 Toggle ; MEM1 %value% ENDON ON Switch1#Boot!=%mem1% DO Backlog Power1 Toggle ; MEM1 %value% ENDON

Thank you all for your help.

BTW
Although the issue seems to be solved, I still think there specific PowerOnState option for this behavior should be considered.

BTW2
I noticed that if the switch changed its state without power, the toggle occurs ~0.1 after boot, so I hope the toggle will be faster than the short (if exist). I don't plan to test it with short willingly.... I think one workaround is to use "PowerOnState 0", and using some rule for emulating the "PowerOnState 3" behavior, after some 0.5 delay. The trick is that it should be consistent with above defined rule. not sure how to write it. - again dedicated command for delay when power is back (regardless PowerOnState value) can solve this, and should be considered.

**Reply by ascillato** [reply]
> Although the issue seems to be solved, I still think there specific PowerOnState option for this behavior should be considered.

Thanks but your proposal have a lot of variables to make just a new simple PowerOnState option. It has several customizations that are linked to your use case, such as which switch is going to be used to toggle (or turn ON or turn OFF) a predefined output while the device is offline, etc. etc. It is not simple and it is better to be addressed by the use of rules like done here. Rules are intended exactly for this type of required behaviors. Besides, this thread is now left here for anyone interested, with 3 different and valid examples for **PowerOnState Switch-Dependant Behavior**.

> I noticed that if the switch changed its state without power, the toggle occurs ~0.1 after boot

Some GPIOs change state at boot time before Tasmota Starts. That is a hardware limitation and can't be addressed by software, sorry. It is a known issue of the ESP8266 chip.
